e2e/internal/keyexec/keyexec.go

```diff
@@ -0,0 +1,347 @@
+func ExportPrivateKey(t *testing.T, kpath, num string, armor bool) (string, []byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	c, err := expect.NewConsole()
+	if err != nil {
+		t.Fatal("Unable to start new console: ", err)
+	}
+	defer c.Close()
+
+	exportCmd := []string{"key", "export", "--secret"}
+
+	if armor {
+		exportCmd = append(exportCmd, "--armor")
+	}
+
+	exportCmd = append(exportCmd, kpath)
+
+	outErr := bytes.NewBuffer(nil)
+
+	cmd := exec.Command(testenv.CmdPath, exportCmd...)
+	cmd.Stdin = c.Tty()
+
+	cmd.Stderr = outErr
+	cmd.Stdout = outErr
+
+	go func() {
+		c.ExpectEOF()
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("unable to run command: %v", err)
+	}
+
+	c.Send(num)
+	c.Send("e2etests\n")
+
+	err = cmd.Wait()
+	cm := fmt.Sprintf("%s %s", testenv.CmdPath, strings.Join(exportCmd, " "))
+
+	return cm, outErr.Bytes(), err
+}
+
+// RunKeyCmd will run a 'singularty key' command, with any args that are set in commands.
+func RunKeyCmd(t *testing.T, cmdPath string, commands []string, file, stdin string) (string, []byte, error) {
+	argv := []string{"key"}
+	argv = append(argv, commands...)
+
+	if file != "" {
+		argv = append(argv, file)
+	}
+
+	cmd := fmt.Sprintf("%s %s", cmdPath, strings.Join(argv, " "))
+	execKey := exec.Command(cmdPath, argv...)
+
+	stdinRun, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	defer stdinRun.Close()
+
+	_, err = io.WriteString(stdinRun, stdin)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	_, err = stdinRun.Seek(0, os.SEEK_SET)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	execKey.Stdin = stdinRun
+	out, err := execKey.CombinedOutput()
+
+	return cmd, out, err
+}
+
+// QuickTestExportImportKey will export a private, and public key (0), and then import them. This is used after
+// generating a newpair.
+func QuickTestExportImportKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	tmpTestDir := filepath.Join(testenv.TestDir, "quick_test_key_verify")
+
+	tests := []struct {
+		name    string
+		private bool
+		armor   bool
+		succeed bool
+	}{
+		{
+			name:    "quick test public",
+			private: false,
+			armor:   false,
+			succeed: true,
+		},
+		{
+			name:    "quick test public armor",
+			private: false,
+			armor:   true,
+			succeed: true,
+		},
+		{
+			name:    "quick test private",
+			private: true,
+			armor:   false,
+			succeed: true,
+		},
+		{
+			name:    "quick test private armor",
+			private: true,
+			armor:   true,
+			succeed: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, test.WithoutPrivilege(func(t *testing.T) {
+			os.RemoveAll(tmpTestDir)
+			os.MkdirAll(tmpTestDir, os.ModePerm)
+			var c string
+			var b []byte
+			var err error
+
+			if tt.private {
+				c, b, err = ExportPrivateKey(t, filepath.Join(tmpTestDir, "export_key.asc"), "0\n", tt.armor)
+			} else {
+				c, b, err = RunKeyCmd(t, testenv.CmdPath, []string{"export"}, filepath.Join(tmpTestDir, "export_key.asc"), "0\n")
```

Comment:
what's the 0 entry?

the bit that is confusing is what does the keyring contain when this is running and why?

Reply:
Key 0 is the newly created key by `newpair`. The keyring is cleared before generating a newpair. This function just tests that key from `newpair` can do what its supposed to.

Reply:
please add a comment indicating that the keyring is empty before running this and that the newly added key gets to be entry 0 in all cases. Also, I believe there's an issue for the keys being zero-based, no?

Reply:
Okay, I changed this test to export all the keys first, then try to import them, one by one. This way the keyring _could_ still have another entity in it.